### PR TITLE
Remove mutating global state in node-exporter objects

### PR DESCRIPTION
**Pull request by paulfantom**
Experiment to encapsulate node-exporter jsonnet definitions and prevent mutating global state.

First step towards https://github.com/prometheus-operator/kube-prometheus/issues/566

### CODE REVIEW

File: jsonnet/kube-prometheus/node-exporter/node-exporter.libsonnet
Comment (brancz) [inline jsonnet/kube-prometheus/node-exporter/node-exporter.libsonnet:4]: I like to use the `error "must provide namespace"` approach, which fails to generate if the config is not set, and forces the caller to configure it.

**Review by brancz — COMMENTED** [review]
Nice start!

**Comment by paulfantom** [comment]
I think this is ready for the next pass of reviews (cc @brancz @metalmatze @kakkoyun). Changes are done:
- moved `kube-rbac-proxy/container.libsonnet` to `kube-rbac-proxy/containerMixin.libsonnet` for now. End goal is to remove this completely and replace with what is new `kube-rbac-proxy/container.libsonnet`
- complete rewrite of node-exporter.libsonnet
- fixed incorrect version number in `app.kubernetes.io/version` label

I still need to fix unit tests and manifest generation, but from local testing it seems that manifests generated from `example.jsonnet` didn't change.

**Comment by paulfantom** [comment]
Rebased to squash commits in some meaningful way.

File: jsonnet/kube-prometheus/kube-prometheus.libsonnet
Comment (brancz) [inline jsonnet/kube-prometheus/kube-prometheus.libsonnet:24]: if we're already refactoring, I would suggest to just have this be plain `image`, we tried this kind of "abstraction" in many ways, but we've learned that the image should just be treated as an opaque string

**Review by brancz — COMMENTED** [review]
So clean :bath: :shower: 

I'm loving that there is almost 0 diff in the generated manifests.

lgtm :+1: 

I wonder whether we should do these refactorings into a branch and merge that branch when everything is done, or cut a release branch now, and just merge the refactorings into master. I would think the later is probably more effective in moving forward faster, but the first would be safer in case it gets abandoned. Your call I would say.

File: jsonnet/kube-prometheus/kube-prometheus.libsonnet
Comment (paulfantom) [reply jsonnet/kube-prometheus/kube-prometheus.libsonnet:24]: I thought about it and applied such approach in new kube-rbac-proxy function. However in case of node-exporter I left it as is due to `app.kubernetes.io/version` label. I can see 3 options of satisfying label and image string:
1) current solution - compose `image` inside a function and pass `version` directly to label.
2) pass `image` string to a function and extract `version` from that string
3) pass `image` and `version` as strings.
I think 3rd one is the most fool-proof, but at the same time not so clean. That said I don't have any preference.

**Comment by paulfantom** [comment]
> I wonder whether we should do these refactorings into a branch and merge that branch when everything is done, or cut a release branch now, and just merge the refactorings into master

We do have `release-0.7` cut recently so I don't think we need another one and merging to `master` should be fine.

File: jsonnet/kube-prometheus/kube-prometheus.libsonnet
Comment (brancz) [reply jsonnet/kube-prometheus/kube-prometheus.libsonnet:24]: I prefer 3, along the lines of, "prefer duplication over the wrong abstraction".

Comment (paulfantom) [reply jsonnet/kube-prometheus/kube-prometheus.libsonnet:24]: Tested the 3rd approach in https://github.com/prometheus-operator/kube-prometheus/pull/862/commits/47f81f46f71a8cb672978bd17242b673472d8193#diff-46fb519eb228dee2044be788ed466089f6652ca4ea439f6f401dd3b509d12dcaR29-R30 and it looks a bit better and cleaner so I will change it here too.

**Comment by paulfantom** [comment]
I added info to readme and changed how `image` is passed.

**Comment by paulfantom** [comment]
And accidentally closed my PR :smile:

File: README.md
Comment (lilic) [inline README.md:23]: The problem is a lot of our users already use master, and we encouraged them to do that in many releases we said this. I suscept this will cause problems, is there no way to avoid it?

Comment (lilic) [reply README.md:23]: Sorry missed the above comments, all for moving fast but might be worth notifying people in a way before this is merged for them to switch from master to latest release? WDYT of cutting a new release and mentioning that in the release notes? Or how long will this state take you think?

Comment (paulfantom) [reply README.md:23]: `release-0.7` is very close to where current `master` is so I don't think a new branch is needed (the diff is in https://github.com/prometheus-operator/kube-prometheus/compare/release-0.7). As for announcing it, it was just done in prometheus community call, this PR mentions it at the top of the readme, and after it is merged I will announce it on slack.

I hope this all will take about a week (depending on how fast reviews will be), there are already 3 pending PRs on top of this one and I have more code on my PC to submit in the next PRs.

Comment (brancz) [reply README.md:23]: I'd say additionally to this statement we can announce it in all the channels on kubernetes slack that are related to kube-prometheus. That, plus a commitment to get it done quickly, sounds like a plan to me.

Comment (lilic) [reply README.md:23]: Sounds good to me, feel free to ping us on getting the reviews quickly in to not disrupt master for folks, thanks!

**Review by lilic — APPROVED** [review]
Thanks! lgtm

**Pull request merged**
